### ISSUE TITLE
Download Station should delete files after download

### ISSUE THREAD

**Issue by demonbane** [Sonarr/Sonarr#1764]
Download Station automatically creates a hard link internally before marking a download as complete, so having Sonarr keep the episodes around on disk to enable seeding is unnecessary. Getting to the actual downloaded files through the filesystem is a giant chore, so the only relevant way to remove downloads is through the Download Station interface anyway.

But since Sonarr wants to keep the files around that's, at best, one more thing to have to remember to clean up and, at worst, doubling your disk usage (for users who don't know to go into Media Management -> Advanced settings -> Use Hardlinks instead of Copy).

**Comment by markus101** [comment]
This is expected behaviour for seeding torrents and is explained in the FAQ. If Download Station supports stopping torrent when seeding goals are met then Sonarr would be able to remove the torrents once they are in a stopped/paused state.

**Comment by demonbane** [comment]
What I'm saying is that Download Station doesn't need the files to be kept around in order to continue seeding. It has an internal copy that is virtually untouchable, so the copy that Sonarr imports can be safely deleted from disk and DS will keep right on seeding.

**Comment by markus101** [comment]
They have other files that they seed from? Do you know the paths they seed from and the path they put completed downloads for Sonarr to import?

**Comment by demonbane** [comment]
The completed downloads are put into a user-configurable directory. In my case, they go into `/<volume_name>/media/incoming`.

The "safe" copies kept around for seeding are under `/<volume_name>/@download/<id>/<torrent name>`. I don't know how the id is determined (looks like it's probably just sequential).

Note: I think that by default Synology uses `volume1` as the default volume name, but that can be changed by the user so it shouldn't be hard-coded.

**Comment by demonbane** [comment]
Here's an example of a torrent that just completed and is now seeding.

First, the obscure internal directory that DownloadStation uses:
```
root@DiskStation:/volume1/@download/757/xxxxxx# ls
total 6.6G
drwxrwxrwx 3 root            root            4.0K Mar 17 15:20 .
drwxrwxrwx 5 DownloadStation DownloadStation 4.0K Mar 17 15:20 ..
-rw-rw-r-- 2 xxxxxxx         torrents          30 Mar 17 14:58 xxxxxx
-rw-rw-r-- 2 xxxxxxx         torrents        4.2K Mar 17 14:58 xxxxxx
-rw-rw-r-- 2 xxxxxxx         torrents        6.6G Mar 17 15:20 xxxxxx
-rw-rw-r-- 2 xxxxxxx         torrents        4.3K Mar 17 14:58 xxxxxx
```

And now the directory that DownloadStation places the files into once they've been fully downloaded (for a movie in this case, waiting for CouchPotato to run the import scan):
```
root@DiskStation:/volume1/media/incoming/movies/xxxxxx# ls
total 6.6G
drwxrwsr-x   2 xxxxxxx torrents  12K Mar 17 15:22 .
drwxrwsr-x   4 xxxxxxx torrents  12K Mar 17 15:22 ..
-rw-rw-r--   2 xxxxxxx torrents   30 Mar 17 14:58 xxxxxx
-rw-rw-r--   2 xxxxxxx torrents 4.2K Mar 17 14:58 xxxxxx
-rw-rw-r--   2 xxxxxxx torrents 6.6G Mar 17 15:20 xxxxxx
-rw-rw-r--   2 xxxxxxx torrents 4.3K Mar 17 14:58 xxxxxx
```

Note the link counts on the files.

**Comment by Taloth** [comment]
(@margaale fyi)

Problem with this is that if we set IsReadOnly to false, then Sonarr will not only move the files, but it'll tell DS to remove the torrent, which you don't want.
Atm I don't see us changing that behavior since it's a little bit more 'involved' than just flicking a switch.

The consequence is that files will remain in the completed folder, but that doesn't really hurt, since it's a hardlink... 3 of em actually, coz Sonarr hardlinks too, provided they're on the same volume.
It doesn't look clean, but I'm inclined to close the issue as won't fix.

**Comment by demonbane** [comment]
It sounds like IsReadOnly is maybe tied to the "Completed Download Handling -> Remove" option? What I'm thinking of is the switch in the blackhole provider that seems to be specific to that provider to just move files instead of copying/hard-linking them. (I checked all the other nzb & torrent providers and none of the other ones share that same setting)

**Comment by margaale** [comment]
The links are kept until torrent is removed. I'm not sure I'm understanding the issue here.

**Comment by demonbane** [comment]
@margaale The problem is with the links in the output folder. Even after the torrent is removed they stay there, and since Sonarr doesn't remove them either, they just stay there until you manually go in and remove them all.

**Comment by margaale** [comment]
In my diskstation the only files in that hidden directory are the ones that are being seed. When sonarr removes the torrent the only files that are kept are the ones imported to my series folder.

**Comment by demonbane** [comment]
Do you have "Auto Extract" enabled in your DownloadStation maybe? On my system, when a torrent is seeding I have 3 hard linked copies. One in my series folder (`/volume1/media/tv/<show_name>`), one in my incoming folder (`/volume1/media/incoming/<torrent_name>`) and the "hidden" copy in `/volume1/@download/<id>/<torrent_name>`.

Once seeding is complete and I remove the torrent from DownloadStation, the hard links that remain are in `/volume1/media/tv/<show_name>` **and** `/volume1/media/incoming/<torrent_name>`. What I'm trying to get rid of is the one in `/volume1/media/incoming`.

**Comment by margaale** [comment]
I don't have autoextract enabled. Do you remove the torrent by hand? Or does sonarr remove it for you? When seeding I also have the three links you mentioned, but when sonarr removes the torrent after being marked as completed by DownloadStation the only file I see is in the series folder. I do have both options under complete download handling set to yes.

**Comment by margaale** [comment]
For example, this is the content of my "hidden" directory: 

```
admin@DiskStation:/volume1/@download$ ll
total 1812
drwxr-xr-x  8 DownloadStation DownloadStation    4096 Mar 24 07:13 .
drwxr-xr-x 36 root            root               4096 Mar 20 11:28 ..
drwxrwxrwx  3 DownloadStation DownloadStation    4096 Mar 19 02:38 723
drwxrwxrwx  3 DownloadStation DownloadStation    4096 Mar 19 03:26 724
drwxrwxrwx  3 DownloadStation DownloadStation    4096 Mar 18 23:42 725
drwxrwxrwx  2 DownloadStation DownloadStation    4096 Jan 23  2014 aMuleTmp
-rw-r--r--  1 DownloadStation DownloadStation 1687929 Dec 26  2013 known2_64.met
-rw-r--r--  1 DownloadStation DownloadStation   15415 Jul  9  2013 known.met
drwxr-xr-x  2 DownloadStation DownloadStation    4096 Oct 14  2015 pyload
drwxr-xr-x  2 DownloadStation DownloadStation    4096 Jul 29  2011 torrents
-rw-r--r--  1 DownloadStation DownloadStation  106453 Mar  9 06:49 transmissiond
```

and this is what is inside one of those ids:

```
admin@DiskStation:/volume1/@download/723$ ll
total 428
drwxrwxrwx 3 DownloadStation DownloadStation   4096 Mar 19 02:38 .
drwxr-xr-x 8 DownloadStation DownloadStation   4096 Mar 24 07:13 ..
-rw------- 1 root            root            185892 Mar 19 02:38 723.resume
-rwxr-xr-x 1 root            root            236733 Mar 18 21:25 723.torrent
drwxrwxrwx 2 root            root              4096 Mar 19 00:39 DOWNLOAD_TR_TMP

```
And this is what i've got right now in downloadstation:

![image](https://cloud.githubusercontent.com/assets/20794934/24298976/753b6d1c-1086-11e7-8168-1f4fff653bfb.png)

**Comment by demonbane** [comment]
The hidden directory is managed by DownloadStation itself and it keeps things around as long as they're "needed". Once a torrent is finished seeding, the contents of the torrent will be removed, but the "housekeeping" files will remain in the directory (the `.resume` and `.torrent` files for example). Once you delete the torrent from DownloadStation, it'll get rid of the directory as well (along with the housekeeping files).

But it sounds like the issue comes down to the "Remove" setting under "Completed Download Handling". I have that set to "No" since I like to go back through my DownloadStation and Sabnzbd, and review what all has downloaded over the last few days occasionally.

So what I'm looking for is the equivalent of the setting in the blackhole provider that would allow me to tell Sonarr to move the files rather than link them, but without telling DownloadStation to remove the download. Is that something that's feasible?

**Comment by margaale** [comment]
@Taloth any comments about his request?

**Comment by Taloth** [comment]
Of course we could change the behavior, in theory (which would mean splitting up IsReadOnly for all download clients), but I prefer not to because it works properly in the normal workflow.
Deleting the torrent manually from DownloadStation is what's causing the problem in combination with having Remove turned off.

The argument about Blackhole isn't really comparable because blackhole has no active-torrent tracking at all, and no remove either.

I'm not totally opposed to changing the behavior, but atm I don't think it's worth the effort.
@markus101 Splitting up IsReadOnly into two booleans, one for import transfertype, one to hold off Remove. Should we do it?

@demonbane  Btw. you can check Activity->History too. It should show everything that Sonarr has grabbed and imported.

**Comment by margaale** [comment]
@Taloth if you don't want to do it because of the extra work maybe I could, if you all are OK with it. I would love to get involved a little deeper.

**Comment by markus101** [comment]
Yeah I don't really see a way around it in this scenario (without ignoring the problem anyways). I don't think it's a priority, but if @margaale wants to pick it up I have no objections.

**Comment by demonbane** [comment]
Just wanted to let you know that I tested this today and it works as planned now. Thank you so much for taking the time to do this @margaale!
